adapters/repos/db/vector/ssdhelpers/product_quantization.go

```diff
@@ -223,9 +224,23 @@ func NewProductQuantizerWithEncoders(cfg ent.PQConfig, distance distancer.Provid
 	}
 
 	pq.kms = encoders
+	pq.buildGlobalDistances()
 	return pq, nil
 }
 
+func (pq *ProductQuantizer) buildGlobalDistances() {
+	pq.globalDistances = make([]float32, pq.ks*pq.ks)
+	for segment := 0; segment < pq.m; segment++ {
+		for i := 0; i < pq.ks; i++ {
+			for j := 0; j < pq.ks; j++ {
+				cX := pq.kms[segment].Centroid(byte(i))
+				cY := pq.kms[segment].Centroid(byte(j))
+				pq.globalDistances[i*pq.ks+j] = pq.distance.Step(cX, cY)
```

Comment:
if step(cx, cy) == step(cy, cx) we could performed calculation only for half of the matrix, as it is symmetrical diagonally. So loop could look like (pseudocode)
```
for i = 0; i < max; i++ {
  for j = 0; j <= i; j++ {
    dist[i][j] = dist[j][i] = calculate(i,j)
  }
}
```

Reply:
Good point!